File: tests/test_guardrails/test_no_hardcoded_values.py
"""
Guardrail Tests: No Hardcoded Values in Code

These tests scan Python and R files for banned patterns that indicate
hardcoded values instead of loading from YAML/DB.

TDD Approach: These tests will FAIL initially, then we fix the code.
"""

import re
from pathlib import Path

import pytest

# Project root
PROJECT_ROOT = Path(__file__).parent.parent.parent


# =============================================================================
# BANNED PATTERNS
# =============================================================================

# Hardcoded AUROC values (e.g., auroc = 0.9110)
AUROC_PATTERN = re.compile(r"auroc\s*[=:]\s*0\.\d{3,}", re.IGNORECASE)

# Hardcoded method names that should come from registry
METHOD_NAME_PATTERNS = [
    re.compile(r'["\']pupil-gt["\']'),
    re.compile(r'["\']MOMENT-gt-finetune["\']'),
    re.compile(r'["\']MOMENT-gt-zeroshot["\']'),
    re.compile(r'["\']ensemble-LOF-MOMENT'),
    re.compile(r'["\']ensembleThresholded-'),
    re.compile(r'["\']UniTS-'),
    re.compile(r'["\']TimesNet-'),
    re.compile(r'["\']OneClassSVM["\']'),
    re.compile(r'["\']LOF["\']'),
    re.compile(r'["\']SAITS["\']'),
    re.compile(r'["\']CSDI["\']'),
]

# Raw hex colors (outside of designated color definition files)
HEX_COLOR_PATTERN = re.compile(r'["\']#[0-9A-Fa-f]{6}["\']')

# Files that ARE allowed to have these patterns
ALLOWED_FILES = {
    # Config files that DEFINE the values
    "colors.yaml",
    "plot_hyperparam_combos.yaml",
    "display_names.yaml",
    "figure_layouts.yaml",
    # Registry files
    "outlier_methods.yaml",
    "imputation_methods.yaml",
    "classifiers.yaml",
    # Test files (can have expected values)
    "test_",
    "conftest.py",
    # This test file itself
    "test_no_hardcoded_values.py",
    # Database extraction (needs to reference method names for mapping)
    "extract_all_configs_to_duckdb.py",
    # Data IO modules that define the registry
    "registry.py",
    "display_names.py",
}


def is_allowed_file(filepath: Path) -> bool:
    """Check if file is allowed to contain banned patterns."""
    name = filepath.name
    for allowed in ALLOWED_FILES:
        if allowed in name:
            return True
    return False


def is_in_comment(line: str, match_start: int) -> bool:
    """Check if match position is inside a comment."""
    # Python comment
    hash_pos = line.find("#")
    if hash_pos != -1 and hash_pos < match_start:
        return True
    # R comment (same as Python)
    return False


# =============================================================================
# TESTS
# =============================================================================


class TestNoHardcodedAUROC:
    """Test that AUROC values are not hardcoded in code."""

    def get_python_files(self):
        """Get all Python source files (excluding tests and configs)."""
        files = []
        for py_file in (PROJECT_ROOT / "src").rglob("*.py"):
            if not is_allowed_file(py_file):
                files.append(py_file)
        return files

    def test_no_hardcoded_auroc_in_python_src(self):
        """Scan src/ Python files for hardcoded AUROC values."""
        violations = []

        for py_file in self.get_python_files():
            content = py_file.read_text()
            lines = content.split("\n")

            for line_num, line in enumerate(lines, 1):
                match = AUROC_PATTERN.search(line)
                if match and not is_in_comment(line, match.start()):
                    violations.append(
                        {
                            "file": str(py_file.relative_to(PROJECT_ROOT)),
                            "line": line_num,
                            "content": line.strip()[:80],
                        }
                    )

        if violations:
            msg = "GUARDRAIL VIOLATION: Hardcoded AUROC values found!\n\n"
            for v in violations:
                msg += f"  {v['file']}:{v['line']}\n    {v['content']}\n\n"
            msg += "FIX: Load AUROC from YAML config or database query."
            pytest.fail(msg)

    def test_no_hardcoded_auroc_in_scripts(self):
        """Scan scripts/ Python files for hardcoded AUROC values."""
        violations = []

        for py_file in (PROJECT_ROOT / "scripts").rglob("*.py"):
            if is_allowed_file(py_file):
                continue

            content = py_file.read_text()
            lines = content.split("\n")

            for line_num, line in enumerate(lines, 1):
                match = AUROC_PATTERN.search(line)
                if match and not is_in_comment(line, match.start()):
                    # Allow in SQL queries (they compute AUROC, not hardcode)
                    if "SELECT" in line.upper() or "ORDER BY" in line.upper():
                        continue
                    violations.append(
                        {
                            "file": str(py_file.relative_to(PROJECT_ROOT)),
                            "line": line_num,
                            "content": line.strip()[:80],
                        }
                    )

        if violations:
            msg = "GUARDRAIL VIOLATION: Hardcoded AUROC values in scripts!\n\n"
            for v in violations:
                msg += f"  {v['file']}:{v['line']}\n    {v['content']}\n\n"
            pytest.fail(msg)


def is_allowed_context(line: str) -> bool:
    """Check if line is in an allowed context for method names.

    Allowed contexts:
    - Docstrings (triple quotes)
    - Examples (e.g., ...)
    - SQL queries (WHERE, IN, SELECT)
    - DataFrame filtering (df[df["col"] == "value"])
    - Display name conversion (.replace(...))
    - Type hints / function signatures
    - Default parameters (model_name: str = "...")
    """
    stripped = line.strip()

    # Docstrings and examples
    if stripped.startswith('"""') or stripped.startswith("'''"):
        return True
    if "e.g." in line.lower() or "example" in line.lower():
        return True

    # SQL queries
    sql_keywords = ["WHERE", "AND", "OR", "IN (", "SELECT", "FROM"]
    if any(kw in line.upper() for kw in sql_keywords):
        return True

    # DataFrame filtering (similar to SQL)
    if '== "' in line or "== '" in line:
        return True

    # Display name conversion (.replace() calls or their arguments)
    if ".replace(" in line:
        return True
    # Also allow if line is a quoted string argument (inside a replace call)
    if stripped.startswith('"') and stripped.endswith('",'):
        return True

    # Type hints or method identifier examples in docstrings
    if "Method identifier" in line or "method_name" in line:
        return True

    # Default parameters in function signatures
    if 'str = "' in line or "str = '" in line:
        return True

    return False


class TestNoHardcodedMethodNames:
    """Test that method names are loaded from registry, not hardcoded."""

    def test_no_hardcoded_method_names_in_viz(self):
        """Scan visualization code for hardcoded method names.

        This test is intentionally strict but allows certain contexts:
        - Docstrings and examples
        - SQL WHERE clauses (filtering, not defining behavior)
        - Display name conversion (.replace())

        NOTE: Configuration dicts that define WHICH methods to show
        should use config files, not hardcoded strings.
        """
        violations = []

        for py_file in (PROJECT_ROOT / "src" / "viz").rglob("*.py"):
            if is_allowed_file(py_file):
                continue

            content = py_file.read_text()
            lines = content.split("\n")

            # Track if we're in a docstring
            in_docstring = False

            for line_num, line in enumerate(lines, 1):
                # Track docstring state
                if '"""' in line or "'''" in line:
                    quote_count = line.count('"""') + line.count("'''")
                    if quote_count == 1:
                        in_docstring = not in_docstring
                    # If opening and closing on same line, stay outside
                    elif quote_count >= 2:
                        pass  # Same line open/close

                if in_docstring:
                    continue

                # Skip allowed contexts
                if is_allowed_context(line):
                    continue

                for pattern in METHOD_NAME_PATTERNS:
                    match = pattern.search(line)
                    if match and not is_in_comment(line, match.start()):
                        violations.append(
                            {
                                "file": str(py_file.relative_to(PROJECT_ROOT)),
                                "line": line_num,
                                "pattern": pattern.pattern,
                                "content": line.strip()[:80],
                            }
                        )

        if violations:
            msg = (
                "GUARDRAIL VIOLATION: Hardcoded method names in visualization code!\n\n"
            )
            for v in violations:
                msg += f"  {v['file']}:{v['line']}\n"
                msg += f"    Pattern: {v['pattern']}\n"
                msg += f"    Line: {v['content']}\n\n"
            msg += "FIX: Load method names from configs/mlflow_registry/ or use display_names module."
            pytest.fail(msg)


class TestNoRawHexColors:
    """Test that colors reference color_definitions, not raw hex."""

    def test_no_raw_hex_in_combo_configs(self):
        """Check that shap_figure_combos uses color_ref, not raw hex."""
        import yaml

        yaml_path = (
            PROJECT_ROOT / "configs" / "VISUALIZATION" / "plot_hyperparam_combos.yaml"
        )
        if not yaml_path.exists():
            pytest.skip("plot_hyperparam_combos.yaml not found")

        config = yaml.safe_load(yaml_path.read_text())

        violations = []

        # Check shap_figure_combos
        shap_combos = config.get("shap_figure_combos", {}).get("configs", [])
        for combo in shap_combos:
            combo_id = combo.get("id", "unknown")
            if "color" in combo:
                color_val = combo["color"]
                if isinstance(color_val, str) and color_val.startswith("#"):
                    violations.append(
                        {
                            "section": "shap_figure_combos",
                            "combo_id": combo_id,
                            "color": color_val,
                        }
                    )

        if violations:
            msg = "GUARDRAIL VIOLATION: Raw hex colors in combo configs!\n\n"
            for v in violations:
                msg += f"  Section: {v['section']}, Combo: {v['combo_id']}\n"
                msg += f'    Has: color: "{v["color"]}"\n'
                msg += '    Should be: color_ref: "--color-xxx"\n\n'
            msg += (
                "FIX: Replace 'color' with 'color_ref' referencing color_definitions."
            )
            pytest.fail(msg)

    def test_no_raw_hex_in_python_viz(self):
        """Check visualization Python files don't have raw hex colors."""
        violations = []

        # Specific files to check (not all - some legitimately define colors)
        files_to_check = [
            PROJECT_ROOT / "src" / "viz" / "featurization_comparison.py",
            PROJECT_ROOT / "src" / "viz" / "foundation_model_dashboard.py",
        ]

        for py_file in files_to_check:
            if not py_file.exists():
                continue

            content = py_file.read_text()
            lines = content.split("\n")

            for line_num, line in enumerate(lines, 1):
                match = HEX_COLOR_PATTERN.search(line)
                if match and not is_in_comment(line, match.start()):
                    # Allow if it's loading from COLORS dict
                    if "COLORS[" in line or "COLORS.get(" in line:
                        continue
                    violations.append(
                        {
                            "file": str(py_file.relative_to(PROJECT_ROOT)),
                            "line": line_num,
                            "content": line.strip()[:80],
                        }
                    )

        if violations:
            msg = "GUARDRAIL VIOLATION: Raw hex colors in visualization code!\n\n"
            for v in violations:
                msg += f"  {v['file']}:{v['line']}\n    {v['content']}\n\n"
            msg += "FIX: Use COLORS dict from plot_config.py or load from YAML."
            pytest.fail(msg)


class TestNoHardcodedInR:
    """Test that R scripts don't have hardcoded values."""

    def test_no_hardcoded_auroc_in_r(self):
        """Scan R figure scripts for hardcoded AUROC values."""
        violations = []

        r_dir = PROJECT_ROOT / "src" / "r" / "figures"
        if not r_dir.exists():
            pytest.skip("R figures directory not found")

        # Pattern for R: auroc = 0.xxx or auroc <- 0.xxx
        r_auroc_pattern = re.compile(r"auroc\s*[=<]-?\s*0\.\d{3,}", re.IGNORECASE)

        for r_file in r_dir.glob("*.R"):
            if "test_" in r_file.name:
                continue

            content = r_file.read_text()
            lines = content.split("\n")

            for line_num, line in enumerate(lines, 1):
                # Skip comments
                stripped = line.strip()
                if stripped.startswith("#"):
                    continue

                match = r_auroc_pattern.search(line)
                if match:
                    violations.append(
                        {
                            "file": str(r_file.relative_to(PROJECT_ROOT)),
                            "line": line_num,
                            "content": line.strip()[:80],
                        }
                    )

        if violations:
            msg = "GUARDRAIL VIOLATION: Hardcoded AUROC values in R scripts!\n\n"
            for v in violations:
                msg += f"  {v['file']}:{v['line']}\n    {v['content']}\n\n"
            msg += "FIX: Load values from JSON file in outputs/r_data/."
            pytest.fail(msg)

    def test_no_hardcoded_numeric_vectors_in_r(self):
        """Check for hardcoded numeric vectors that should come from data."""
        violations = []

        r_dir = PROJECT_ROOT / "src" / "r" / "figures"
        if not r_dir.exists():
            pytest.skip("R figures directory not found")

        # Pattern for suspicious numeric vectors: c(0.xxx, 0.xxx, ...)
        numeric_vector_pattern = re.compile(r"c\(\s*0\.\d{2,}.*0\.\d{2,}.*\)")

        for r_file in r_dir.glob("*.R"):
            if "test_" in r_file.name:
                continue

            content = r_file.read_text()
            lines = content.split("\n")

            for line_num, line in enumerate(lines, 1):
                stripped = line.strip()
                if stripped.startswith("#"):
                    continue

                match = numeric_vector_pattern.search(line)
                if match:
                    # Check if it looks like metric data
                    if any(
                        kw in line.lower()
                        for kw in ["auroc", "brier", "ci_lo", "ci_hi"]
                    ):
                        violations.append(
                            {
                                "file": str(r_file.relative_to(PROJECT_ROOT)),
                                "line": line_num,
                                "content": line.strip()[:100],
                            }
                        )

        if violations:
            msg = "GUARDRAIL VIOLATION: Hardcoded metric vectors in R scripts!\n\n"
            for v in violations:
                msg += f"  {v['file']}:{v['line']}\n    {v['content']}\n\n"
            msg += "FIX: Load data from JSON file, not hardcoded vectors."
            pytest.fail(msg)


class TestExplicitPathSelection:
    """Test that database paths are explicit, not fallback chains."""

    def test_no_fallback_path_chains(self):
        """Check that get_database_path doesn't have multiple fallbacks."""
        plot_config_path = PROJECT_ROOT / "src" / "viz" / "plot_config.py"

        if not plot_config_path.exists():
            pytest.skip("plot_config.py not found")

        content = plot_config_path.read_text()

        # Check for multiple path definitions in a list (indicates fallback)
        if content.count("Path(") > 5 and "possible_paths" in content:
            # Count how many paths are in the fallback list
            fallback_match = re.search(
                r"possible_paths\s*=\s*\[(.*?)\]", content, re.DOTALL
            )
            if fallback_match:
                fallback_content = fallback_match.group(1)
                path_count = fallback_content.count("Path(")
                if path_count > 2:
                    pytest.fail(
                        f"GUARDRAIL VIOLATION: {path_count}-path fallback chain in plot_config.py!\n\n"
                        "Having multiple fallback paths means different environments may load different databases.\n\n"
                        "FIX: Use single canonical path with explicit error if not found."
                    )

File: tests/test_r_guardrails/test_no_hardcoded_values.R
# Test: No Hardcoded Values in R Figure Scripts
# ==============================================
# Guardrail tests ensuring R scripts load from YAML/JSON, not hardcoded values.
#
# Run: Rscript -e "testthat::test_file('tests/test_r_guardrails/test_no_hardcoded_values.R')"
#
# Created: 2026-01-27
# Author: Foundation PLR Team

suppressPackageStartupMessages({
  library(testthat)
})

# Find project root
find_project_root <- function() {
  markers <- c("pyproject.toml", "CLAUDE.md", ".git")
  dir <- getwd()
  while (dir != dirname(dir)) {
    if (any(file.exists(file.path(dir, markers)))) {
      return(dir)
    }
    dir <- dirname(dir)
  }
  stop("Could not find project root")
}

PROJECT_ROOT <- find_project_root()

# ==============================================================================
# TEST: No hardcoded model names in figure scripts
# ==============================================================================

test_that("No hardcoded model names in figure scripts", {
  figure_scripts <- list.files(
    file.path(PROJECT_ROOT, "src/r/figures"),
    pattern = "\\.R$",
    full.names = TRUE
  )

  if (length(figure_scripts) == 0) {
    skip("No R figure scripts found")
  }

  # Patterns that indicate hardcoded model names
  banned_patterns <- c(
    '"pupil-gt"',           # Model name in quotes
    '"MOMENT-gt-finetune"', # Model name
    '"ensemble-LOF',        # Model name
    '"OneClassSVM"',        # Model name
    '"SAITS"',              # Imputation method
    '"CSDI"'                # Imputation method
  )

  violations <- list()

  for (script in figure_scripts) {
    content <- paste(readLines(script, warn = FALSE), collapse = "\n")

    for (pattern in banned_patterns) {
      # Check for pattern outside comments
      lines <- strsplit(content, "\n")[[1]]
      for (i in seq_along(lines)) {
        line <- lines[i]
        # Skip comments
        if (grepl("^\\s*#", line)) next

        if (grepl(pattern, line, fixed = TRUE)) {
          # Allow in certain contexts
          # - String replacement calls
          # - Variable definitions from YAML
          # - Display name mapping
          allowed_contexts <- c("yaml::", "fromJSON", "_display_name", "load_", "%||%")
          if (any(sapply(allowed_contexts, function(ctx) grepl(ctx, line)))) next

          violations[[length(violations) + 1]] <- list(
            file = basename(script),
            line = i,
            pattern = pattern,
            content = substr(trimws(line), 1, 60)
          )
        }
      }
    }
  }

  if (length(violations) > 0) {
    msg <- "GUARDRAIL VIOLATION: Hardcoded model names in R scripts!\n\n"
    for (v in violations) {
      msg <- paste0(msg, sprintf("  %s:%d\n    Pattern: %s\n    Line: %s\n\n",
                                  v$file, v$line, v$pattern, v$content))
    }
    msg <- paste0(msg, "FIX: Load model names from YAML config or display_names.yaml")
    fail(msg)
  }
})

# ==============================================================================
# TEST: No hardcoded color vectors
# ==============================================================================

test_that("No hardcoded color vectors in figure scripts", {
  figure_scripts <- list.files(
    file.path(PROJECT_ROOT, "src/r/figures"),
    pattern = "\\.R$",
    full.names = TRUE
  )

  if (length(figure_scripts) == 0) {
    skip("No R figure scripts found")
  }

  # Pattern for hardcoded hex color vectors
  color_vector_pattern <- 'c\\("[^"]*#[0-9A-Fa-f]{6}'

  violations <- list()

  for (script in figure_scripts) {
    content <- paste(readLines(script, warn = FALSE), collapse = "\n")
    lines <- strsplit(content, "\n")[[1]]

    for (i in seq_along(lines)) {
      line <- lines[i]
      if (grepl("^\\s*#", line)) next  # Skip comments

      if (grepl(color_vector_pattern, line, perl = TRUE)) {
        violations[[length(violations) + 1]] <- list(
          file = basename(script),
          line = i,
          content = substr(trimws(line), 1, 80)
        )
      }
    }
  }

  if (length(violations) > 0) {
    msg <- "GUARDRAIL VIOLATION: Hardcoded color vectors in R scripts!\n\n"
    for (v in violations) {
      msg <- paste0(msg, sprintf("  %s:%d\n    %s\n\n", v$file, v$line, v$content))
    }
    msg <- paste0(msg, "FIX: Load colors from plot_hyperparam_combos.yaml via config_loader.R")
    fail(msg)
  }
})

# ==============================================================================
# TEST: All figure scripts load config from YAML
# ==============================================================================

test_that("All figure scripts use config loading functions", {
  figure_scripts <- list.files(
    file.path(PROJECT_ROOT, "src/r/figures"),
    pattern = "^fig.*\\.R$",  # Only files starting with "fig"
    full.names = TRUE
  )

  if (length(figure_scripts) == 0) {
    skip("No R figure scripts found")
  }

  # Patterns that indicate proper config loading
  config_patterns <- c(
    "load_figure_config",
    "load_figure_combos",
    "yaml::read_yaml",
    "validate_data_source",
    "load_figure_all"
  )

  missing_config <- c()

  for (script in figure_scripts) {
    content <- paste(readLines(script, warn = FALSE), collapse = "\n")

    has_config_loading <- any(sapply(config_patterns, function(p) grepl(p, content)))

    if (!has_config_loading) {
      missing_config <- c(missing_config, basename(script))
    }
  }

  if (length(missing_config) > 0) {
    msg <- sprintf(
      "GUARDRAIL VIOLATION: %d figure scripts don't load config from YAML:\n\n  %s\n\nFIX: Add config loading using config_loader.R functions.",
      length(missing_config),
      paste(missing_config, collapse = "\n  ")
    )
    # This is a warning, not a failure, since some scripts might have valid reasons
    warning(msg)
  }
})

# ==============================================================================
# TEST: No on-the-fly model selection (e.g., configs[1:4])
# ==============================================================================

test_that("No on-the-fly model selection", {
  figure_scripts <- list.files(
    file.path(PROJECT_ROOT, "src/r/figures"),
    pattern = "\\.R$",
    full.names = TRUE
  )

  if (length(figure_scripts) == 0) {
    skip("No R figure scripts found")
  }

  # Patterns indicating on-the-fly selection
  banned_patterns <- c(
    "configs\\[1:4\\]",        # Hardcoded slice
    "configs\\[1:min",         # Hardcoded slice with min
    "head\\(configs",          # Taking first N
    "configs %>% head",        # Pipe version
    'grepl.*"MOMENT".*filter'  # Pattern-based filtering
  )

  violations <- list()

  for (script in figure_scripts) {
    content <- paste(readLines(script, warn = FALSE), collapse = "\n")

    for (pattern in banned_patterns) {
      if (grepl(pattern, content, perl = TRUE)) {
        violations[[length(violations) + 1]] <- list(
          file = basename(script),
          pattern = pattern
        )
      }
    }
  }

  if (length(violations) > 0) {
    msg <- "GUARDRAIL VIOLATION: On-the-fly model selection detected!\n\n"
    for (v in violations) {
      msg <- paste0(msg, sprintf("  %s: uses pattern '%s'\n", v$file, v$pattern))
    }
    msg <- paste0(msg, "\nFIX: Define which configs to use in YAML, not in R code.")
    fail(msg)
  }
})

# ==============================================================================
# TEST: case_when not used for display names
# ==============================================================================

test_that("case_when not used for hardcoded display names", {
  figure_scripts <- list.files(
    file.path(PROJECT_ROOT, "src/r/figures"),
    pattern = "\\.R$",
    full.names = TRUE
  )

  if (length(figure_scripts) == 0) {
    skip("No R figure scripts found")
  }

  # Pattern for case_when with hardcoded strings (display name conversion)
  # This catches: case_when(method == "pupil-gt" ~ "Ground Truth", ...)
  case_when_pattern <- 'case_when.*~.*"[A-Z]'

  violations <- list()

  for (script in figure_scripts) {
    content <- paste(readLines(script, warn = FALSE), collapse = "\n")
    lines <- strsplit(content, "\n")[[1]]

    for (i in seq_along(lines)) {
      line <- lines[i]
      if (grepl("^\\s*#", line)) next

      if (grepl(case_when_pattern, line, perl = TRUE)) {
        # Check if it's for display name conversion
        if (grepl("display|label|name", line, ignore.case = TRUE)) {
          violations[[length(violations) + 1]] <- list(
            file = basename(script),
            line = i,
            content = substr(trimws(line), 1, 80)
          )
        }
      }
    }
  }

  if (length(violations) > 0) {
    msg <- "GUARDRAIL VIOLATION: Using case_when for display name conversion!\n\n"
    for (v in violations) {
      msg <- paste0(msg, sprintf("  %s:%d\n    %s\n\n", v$file, v$line, v$content))
    }
    msg <- paste0(msg, "FIX: Use apply_display_names() from common.R instead.")
    fail(msg)
  }
})

message("\n[test_no_hardcoded_values.R] All guardrail tests completed.")
